Thread safety analysis: Fix a bug in handling temporary constructors

ziqingluo-90 opened the pull request:
The commit 54bfd04846156dbd5e0a6b88f539c3d4569a455f introduces general handling of constructor and destructor calls.  Consider a pair of TEMPORARY constructor and destructor calls of a "SCOPED_CAPABILITY" object.  At the constructor call, a resource representing the object itself is added to the state.  But since it is a temporary object, the analyzer may not be able to find an AST part to identify the resource (As opposed to the normal case of using `VarDecl` as an identifier). Later at the destructor call, the resource needs to be removed from the state.  Since the resource cannot be identified by anything other than itself, the analyzer maintains a map from constructor calls to their created resources so that the analyzer could get the proper resource for removal at the destructor call via look up the map.  The issue is that the map lives within a CFG block.  That means in case the pair of temporary constructor and destructor calls are not in one CFG block, the analyzer is not able to remove the proper resource at the destructor call.  Although the two calls stay in one CFG block usually, they separate in the following example:

```
if (f(MutexLock{lock}) || x) { // <-- the temporary Ctor and Dtor of MutexLock are in different CFG blocks

}
```

The fix is simple. It extends the life of the map to that of the entire CFG.

aaronpuchert commented on clang/lib/Analysis/ThreadSafety.cpp:
```diff
@@ -1544,7 +1544,10 @@ class BuildLockset : public ConstStmtVisitor<BuildLockset> {
   const FactSet &FunctionExitFSet;
   /// Maps constructed objects to `this` placeholder prior to initialization.
-  llvm::SmallDenseMap<const Expr *, til::LiteralPtr *> ConstructedObjects;
+  /// The map lives longer than this builder so this is a reference to an
+  /// existing one. The map lives so long as the CFG while this builder is for
+  /// one CFG block.
```
It's probably sufficient to explain this in the commit message.

aaronpuchert commented on clang/lib/Analysis/ThreadSafety.cpp:
```diff
@@ -2392,6 +2397,8 @@ void ThreadSafetyAnalyzer::runAnalysis(AnalysisDeclContext &AC) {
   for (const auto &Lock : LocksReleased)
     ExpectedFunctionExitSet.removeLock(FactMan, Lock);
 
+  ConstructedObjectMapTy ConstructedObjects;
```
I wonder if we should put it into the `ThreadSafetyAnalyzer`, where we already have the similar `LocalVariableMap`. Then we don't need to pass it into `BuildLockset`, because that already has a pointer to the `ThreadSafetyAnalyzer`.

aaronpuchert commented on clang/test/SemaCXX/warn-thread-safety-analysis.cpp:
```diff
@@ -1718,6 +1720,13 @@ struct TestScopedLockable {
   }
 
+  void temporary2(int x) {
+    if (check(MutexLock{&mu1}) || x) {
+
+    }
```
The `if` is probably not needed here, right? We could just write
```suggestion
    check(MutexLock{&mu1}) || x;
```
and should have the same CFG artifact.

aaronpuchert commented on clang/test/SemaCXX/warn-thread-safety-analysis.cpp:
```diff
@@ -1718,6 +1720,13 @@ struct TestScopedLockable {
+    if (check(MutexLock{&mu1}) || x) {
+
+    }
+    check(MutexLock{&mu1});
```
The `check` here doesn't do anything, right?

aaronpuchert commented on clang/test/SemaCXX/warn-thread-safety-analysis.cpp:
```diff
@@ -1718,6 +1720,13 @@ struct TestScopedLockable {
     MutexLock{&mu1}, a = 5;
   }
 
+  void temporary2(int x) {
```
I would suggest a speaking name for the test, like `temporary_cfg`.

aaronpuchert reviewed (commented):
I have some suggestions, but in principle this is absolutely right. Thanks for finding this and providing a fix!

>  The issue is that the map lives within a CFG block.

It didn't cross my mind to check how long `BuildLockset` lived, I always assumed it lived for the entire function. But you're right, it does not, and is therefore an unsuitable home for the map.

ziqingluo-90 committed:
Thread safety analysis: Fix a bug in handling temporary constructors

Extends the lifetime of the map `ConstructedObjects` to be of the
whole CFG so that the map can connect temporary Ctor and Dtor in
different CFG blocks.

ziqingluo-90 commented:
@aaronpuchert Thanks for your comment!  I have addressed them.   Do you mind to take an another look and approve it if it looks good to you?

aaronpuchert commented on clang/lib/Analysis/ThreadSafety.cpp:
```diff
@@ -1530,7 +1532,6 @@ void ThreadSafetyAnalyzer::getEdgeLockset(FactSet& Result,
 }
 
 namespace {
-
```
Nitpick: can you undo the whitespace change?

ziqingluo-90 commented on clang/lib/Analysis/ThreadSafety.cpp:
```diff
@@ -2487,15 +2486,15 @@ void ThreadSafetyAnalyzer::runAnalysis(AnalysisDeclContext &AC) {
               LocksetBuilder.handleCall(nullptr, DD, Object->second,
                                         TD.getBindTemporaryExpr()->getEndLoc());
-            LocksetBuilder.ConstructedObjects.erase(Object);
+            LocksetBuilder.Analyzer->ConstructedObjects.erase(Object);
```
Done

aaronpuchert reviewed (approved):
Thanks, looks good to me!
